File: purchases/src/test/java/com/revenuecat/purchases/EntitlementTest.java
package com.revenuecat.purchases;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.Map;

import static junit.framework.Assert.assertEquals;

@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE)
public class EntitlementTest {

    private final Entitlement.Factory factory = new Entitlement.Factory();

    Map<String, Entitlement> buildEntitlementMap(String json) throws JSONException {
        JSONObject object = new JSONObject(json);
        return factory.build(object);
    }

    @Test
    public void factoryHandlesEmptyCase() throws JSONException {
        Map<String, Entitlement> entitlementMap = buildEntitlementMap("{}");
        assertEquals(0, entitlementMap.size());
    }

    @Test
    public void factoryHandlesEntitlementNoOfferings() throws JSONException {
        Map<String, Entitlement> entitlementMap = buildEntitlementMap("{'pro': {'offerings': {}}}");
        assertEquals(1, entitlementMap.size());
        Entitlement e = entitlementMap.get("pro");
        Map<String, Offering> offerings = e.getOfferings();
        assertEquals(0, offerings.size());
    }

    @Test
    public void factoryHandlesOneOffering() throws JSONException {
        Map<String, Entitlement> entitlementMap = buildEntitlementMap("{'pro': {'offerings': {'monthly': {'active_product_identifier': 'onemonth_freetrial'}}}}");
        Entitlement e = entitlementMap.get("pro");
        Map<String, Offering> offerings = e.getOfferings();
        assertEquals(1, offerings.size());
        Offering o = offerings.get("monthly");
        assertEquals("onemonth_freetrial", o.getActiveProductIdentifier());
    }

    @Test
    public void factoryHandlesMultipleOfferings() throws JSONException {
        Map<String, Entitlement> entitlementMap = buildEntitlementMap("{'pro': {'offerings': {'monthly': {'active_product_identifier': 'onemonth_freetrial'},'annual': {'active_product_identifier': 'oneyear_freetrial'}}}}");
        Entitlement e = entitlementMap.get("pro");
        Map<String, Offering> offerings = e.getOfferings();

        Offering monthly = offerings.get("monthly");
        Offering annual = offerings.get("annual");

        assertEquals("onemonth_freetrial", monthly.getActiveProductIdentifier());
        assertEquals("oneyear_freetrial", annual.getActiveProductIdentifier());
    }

    @Test
    public void factoryHandlesMultipleEntitlments() throws JSONException {
        Map<String, Entitlement> entitlementMap = buildEntitlementMap("{'pro': {'offerings': {'monthly': {'active_product_identifier': 'onemonth_freetrial'},'annual': {'active_product_identifier': 'oneyear_freetrial'}}}, 'pro2': {'offerings': {'monthly': {'active_product_identifier': 'onemonth_freetrial2'},'annual': {'active_product_identifier': 'oneyear_freetrial2'}}}}");
        assertEquals(2, entitlementMap.size());
    }
}

File: purchases/src/test/java/com/revenuecat/purchases/EntitlementTest.kt
package com.revenuecat.purchases

import android.support.test.runner.AndroidJUnit4
import org.json.JSONObject
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.annotation.Config

import org.assertj.core.api.Assertions.assertThat

@RunWith(AndroidJUnit4::class)
@Config(manifest = Config.NONE)
class EntitlementTest {

    private val factory = Entitlement

    @Test
    fun `given an empty case, factory can handle it`() {
        val entitlementMap = buildEntitlementMap("{}")
        assertThat(entitlementMap).`as`("check entitleMap is empty").isEmpty()
    }

    @Test
    fun `given an entitlement with no offerings, factory can handle it`() {
        val entitlementMap = buildEntitlementMap("{'pro': {'offerings': {}}}")
        assertThat(entitlementMap.size).`as`("check there is one entitle").isEqualTo(1)

        val e = entitlementMap["pro"]
        val offerings = e!!.offerings
        assertThat(offerings).`as`("check there is no offerings").isEmpty()
    }

    @Test
    fun `given an entitlement with one offering, factory can handle it`() {
        val entitlementMap =
            buildEntitlementMap("{'pro': {'offerings': {'monthly': {'active_product_identifier': 'onemonth_freetrial'}}}}")
        val e = entitlementMap["pro"]
        val offerings = e!!.offerings
        assertThat(offerings.size).`as`("check there is one offering").isEqualTo(1)
        val o = offerings["monthly"]
        assertThat(o!!.activeProductIdentifier)
            .`as`("check there is one offering onemonth_freetrial")
            .isEqualTo("onemonth_freetrial")
    }

    @Test
    fun `given an entitlement with multiple offerings, factory can handle it`() {
        val entitlementMap =
            buildEntitlementMap("{'pro': {'offerings': {'monthly': {'active_product_identifier': 'onemonth_freetrial'},'annual': {'active_product_identifier': 'oneyear_freetrial'}}}}")
        val e = entitlementMap["pro"]
        val offerings = e!!.offerings

        val monthly = offerings["monthly"]
        val annual = offerings["annual"]

        assertThat(monthly!!.activeProductIdentifier)
            .`as`("check there is one offering onemonth_freetrial")
            .isEqualTo("onemonth_freetrial")

        assertThat(annual!!.activeProductIdentifier)
            .`as`("check there is one offering oneyear_freetrial")
            .isEqualTo("oneyear_freetrial")
    }

    @Test
    fun `given multiple entitlements with multiple offerings, factory can handle it`() {
        val entitlementMap =
            buildEntitlementMap("{'pro': {'offerings': {'monthly': {'active_product_identifier': 'onemonth_freetrial'},'annual': {'active_product_identifier': 'oneyear_freetrial'}}}, 'pro2': {'offerings': {'monthly': {'active_product_identifier': 'onemonth_freetrial2'},'annual': {'active_product_identifier': 'oneyear_freetrial2'}}}}")
        assertThat(entitlementMap.size).`as`("check there is two entitles").isEqualTo(2)
        assertThat(entitlementMap["pro"]!!.offerings.size).`as`("check there is two pro offerings").isEqualTo(2)
        assertThat(entitlementMap["pro2"]!!.offerings.size).`as`("check there is two pro2 offerings").isEqualTo(2)
    }

    private fun buildEntitlementMap(json: String): Map<String, Entitlement> {
        return factory.build(JSONObject(json))
    }
}
